### ISSUE TITLE
Is there an API for other plugins to use?

### ISSUE THREAD

**Issue by czottmann** [srg-kostyrko/obsidian-journal#78]
Hi, I'm the author of [Actions For Obsidian](https://actions.work/actions-for-obsidian) and its companion plugin [Actions URI](https://zottmann.dev/obsidian-actions-uri/). The latter is currently hooking into the core plugin Daily Notes, the community plugin Periodic Notes, via [obsidian-daily-notes-interface](https://github.com/liamcain/obsidian-daily-notes-interface).

My customers are asking about my app & plugin supporting Journals, but AFAICT, you don't offer a clear API for checking periodic notes or creating them, right?

Thanks!

**Comment by srg-kostyrko** [comment]
yes, right now there is no such api.
I considered adding it in future at some point.
As it looks there is a need for it already - I'll bump it's priority and will start working on it right after v2 release (as it has lots of changes to structure)

please let me know if you have some specific API capabilities in mind

**Comment by czottmann** [comment]
@srg-kostyrko Thanks, I appreciate it! Basically, as a dev writing plugins that might interact with Journals, I'd **love** to see an up-to-date fork of the aforementioned [obsidian-daily-notes-interface](https://github.com/liamcain/obsidian-daily-notes-interface) package that also supports Journals, but I know that this is a stretch 😅

So I'd prefer a npm package _like_ that. Mostly an abstraction that offers a few methods for …

- checking if Journals is installed/active
- returning the current daily/weekly/etc. periodic note, and possibly the most recent (i.e., if today's note doesn't exist, get yesterday's)
- creating a daily/weekly/etc. periodic note

This list is informed by my reduced use cases, of course. Personally, I don't need anything related to calendars or any other types of notes right now. Other devs might need other endpoints, I guess. But overall, I'd prefer it if the API was "just" a standalone npm package, it'd be much easier to handle for other plugin authors like me.

Thank you!

**Comment by czottmann** [comment]
> you can have multiple daily (same for weekly, monthly etc) journals, so there might be several daily notes for same day.

Ah, okay. I was not aware of that, to be honest.

Some more background: I had one or two customers asking for my macOS/iOS app supporting Journals. They use periodic notes, but they are not using the Periodic Note plugin anymore, because they switched to Journals. So I naïvely just figured that Journals follows the same idea (one set of daily/weekly/etc. per vault), but clearly I was wrong! 😅 My app currently does not have a concept of "journals" inside a vault, so there's structural discrepancy between what your plugin is allowing for and what my app does.

Thanks for your explanations! 🙏🏼 I now have a clearer understanding of what I'd need to do to actually add support for Journals to Actions URI (and to [Actions For Obsidian](https://actions.work/actions-for-obsidian)), but I won't tackle that now. **So, I'm good for the time being.**

**Comment by Jnolcox** [comment]
How about a function that lists journals to choose from that have `journal-selection: daily` if there is more than one.

**Comment by srg-kostyrko** [comment]
I would expect just getting note for specific journal as easier approach
but can also add arbitrary filtering - in the end it comes down to how other plugin will use it

**Comment by franciskafieh** [comment]
Hey @srg-kostyrko are there any updates on this? I have been using the daily-notes-interface, but I was made aware that is it out of date and your plugin seems to be the modern replacement.

Thank you!

**Comment by srg-kostyrko** [comment]
This was not on my priority list, as czottmann, who asked for it initially, said he would hold off on integration for now.
I'll up it and add to one of upcoming milestones. No ETA for now.

**Comment by franciskafieh** [comment]
Thank you very much!

**Comment by Jihad** [comment]
@srg-kostyrko @czottmann  Would like to add my voice here. I'd love to have a URI link that opens/creates a journal directly, by it's id for example - or maybe via related "command" id.

Usecase Context:
I use Apple shortcuts and Drafts App a lot, with workflows of scripts that writes directly to .md files (no need to open obsidian) - one of my flows is writing directly under a specific heading in my Daily or "Work Weekly" journals.

But, when they are not created yet, it fails. Have to open obsidian, find command to create it, then re-run the flow.

It would be great to have a URI, that opens obsidian and execute my journal command directly.

**Comment by srg-kostyrko** [comment]
@Jihad from what you discribe you need URI support https://github.com/srg-kostyrko/obsidian-journal/issues/85
this ticket is more about some package that other plugins can use for integration and won't help with your use case

**Comment by Jihad** [comment]
@srg-kostyrko Oh that's right. Thank you 🫡

**Comment by helpimfalling** [comment]
Would also love this to be implemented! Was hoping I could easily use the raycast obsidian plugin with journals but it needs URI support.

**Comment by srg-kostyrko** [comment]
Reopened — this was closed by accident on 14 August, not resolved.

A commit in the v2 → v3 parity audit referred to that audit's own item 78 as `#78` in its message, and GitHub read it as a reference to this issue and closed it when the branch merged. The commit touched nothing but an internal notes file.

There is still no plugin-facing API. Sorry for the noise.

**Comment by srg-kostyrko** [comment]
Released in **3.2.0**. [`obsidian-journals-api`](https://www.npmjs.com/package/obsidian-journals-api) is on npm — the typed surface plus a locator, no domain logic, so it cannot drift from the plugin.

```ts
import { getJournalsApi } from "obsidian-journals-api";

const journals = getJournalsApi(this.app);
if (!journals) return; // not installed, or not enabled
```

Call it at the point of use rather than caching it at load: there is no readiness event, and reloading the plugin replaces the object. Every method is async and waits internally for whatever it needs, so you never have to schedule around Journals' startup.

@czottmann — against your three bullets:

- **checking if Journals is installed/active** — `getJournalsApi(app) !== null`.
- **returning the current daily/weekly/etc. note** — `notesFor(selector, date)`, e.g. `notesFor({ writeType: "day" }, "today")`.
- **creating one** — `ensureNote(selector, date)` creates it if missing and resolves to `{ note, created }`; `openNote` does the same and opens it.

On the several-journals-per-type problem you and @Jnolcox were both circling, both of the approaches discussed shipped rather than one:

- **`JournalSelector`** is either a journal name, or `{ journal?, writeType?, shelf? }` with the fields ANDed — so "the note from this specific journal" and "any daily journal on this shelf" are both expressible. `shelf: null` means journals on no shelf; an empty selector matches everything.
- **Reads fan out.** `notesFor` returns an array — empty, one, or several — each entry carrying its journal info.
- **Writes resolve to one.** `ensureNote` and `openNote` produce a single note, showing the journal picker when the selector matches more than one, exactly as clicking a calendar cell does.

The rest of the surface: `listJournals`, `journalInfo(name)`, `journalOf(file)` for the reverse lookup, and `on(event, handler)` for `journalCreated` / `journalDeleted` / `noteAdded` / `noteRemoved`, returning an unsubscribe disposer.

**One thing you asked for that is not there:** "the most recent — if today's note doesn't exist, get yesterday's". The plugin has that as a command but it is not on the API surface. Worth a separate issue if it is still useful to you.

`apiVersion` is an integer bumped only on a breaking change, and the package major tracks it, so `1.x` describes `apiVersion 1`. Additive changes — including new error codes — never move it, so handle the default case when branching on `error.code`.

Full reference, including a migration table from `obsidian-daily-notes-interface`: [`docs/plugin-api.md`](https://github.com/srg-kostyrko/obsidian-journal/blob/main/docs/plugin-api.md).

@franciskafieh — this is the one you were waiting on.
